### PR TITLE
Some proofs missing from Data.Integer.Properties

### PR DESCRIPTION
**Pull request by gallais**
Not too sure about the names

### CODE REVIEW

**Comment by gallais** [comment]
This one can stay on the backburner for a bit: I'll keep adding things that I find I
am missing in my cleanup of https://github.com/gallais/agda-presburger/tree/refactor

**Comment by gallais** [comment]
I think I now have everything I need. That last one was not necessary but I couldn't
resist the temptation to bump this PR above 1000 additions.

There are still *plenty* of things missing in `Data.Integer.Properties`.

**Comment by MatthewDaggitt** [comment]
Wow, now this is a pull request!

At first glance mind if I remove the `StrictPartialOrderReasoning` module? It's not used in the PR and it should be superseded by #185...

**Comment by gallais** [comment]
It was used in one place in `Data.Integer.Properties` but we can reason using `≤-Reasoning`
instead even though it's not as elegant.

I've added a checkbox to #185's top comment to remember to fix that proof once we have proper
`<-Reasoning`.

**Comment by gallais** [comment]
Strange. Looks like the module already existed before this PR: https://agda.github.io/agda-stdlib/Relation.Binary.StrictPartialOrderReasoning.html

File: src/Data/Nat/Properties.agda [hunk 1]
```diff
@@ -463,6 +466,10 @@ i+j≡0⇒i≡0 (suc i) ()
 i+j≡0⇒j≡0 : ∀ i {j} → i + j ≡ 0 → j ≡ 0
 i+j≡0⇒j≡0 i {j} i+j≡0 = i+j≡0⇒i≡0 j (trans (+-comm j i) (i+j≡0))
 
+m≢0⇒m≡s[pred[m]] : ∀ {m} → m ≢ 0 → m ≡ suc (pred m)
```
Comment (MatthewDaggitt) [inline src/Data/Nat/Properties.agda]: Really nitpicky I know, but usually the complex argument goes on the left hand side of the equality. Also I'd probably call it `m≢0⇒suc[pred[m]]≡m`. Maybe we can even drop `m≢0`, not sure though...

File: src/Data/Integer/Properties.agda [hunk 2]
```diff
@@ -692,18 +787,180 @@ pos-distrib-* (suc x) (suc y) = refl
...
+        = +≤+ z≤n
+
+*-monoˡ-≤-non-neg : ∀ n → (+ n *_) Preserves _≤_ ⟶ _≤_
+*-monoˡ-≤-non-neg n {i} {j} i≤j
+  rewrite *-comm (+ n) i
+        | *-comm (+ n) j
+        = *-monoʳ-≤-non-neg n i≤j
+
+*-monoˡ-≤-pos : ∀ n → (+ suc n *_) Preserves _≤_ ⟶ _≤_
+*-monoˡ-≤-pos n = *-monoˡ-≤-non-neg (suc n)
```
Comment (MatthewDaggitt) [inline src/Data/Integer/Properties.agda]: Do we need this as well as `*-monoˡ-≤-non-neg`?

File: src/Data/Integer/Properties.agda [hunk 3]
```diff
@@ -517,40 +588,49 @@ private
...
         | ⊖-≰ (b≰a ∘ ℕₚ.*-cancelʳ-≤ b a c)
-        | -◃n≡-n ((b ∸ a) ℕ* suc c)
+        | -◃n≡-n ((b ∸ a) ℕ.* suc c)
         | ℕₚ.*-distribʳ-∸ (suc c) b a
         = refl
 
+*-distribˡ-+ : _*_ DistributesOverˡ _+_
+*-distribˡ-+ = comm+distrʳ⇒distrˡ *-comm *-distribʳ-+
+
+sm*n≡n+m*n : ∀ m n → sucℤ m * n ≡ n + m * n
```
Comment (MatthewDaggitt) [inline src/Data/Integer/Properties.agda]: This should probably be called `[1+m]*n≡n+m*n` instead?

File: src/Data/Integer/Properties.agda [hunk 4]
```diff
@@ -296,6 +292,32 @@ distribʳ-⊖-+-neg a b c = begin
...
+  + 0 + m       ≡⟨ +-identityˡ m ⟩
+  m ∎ where open ≡-Reasoning
+
+pred-suc : ∀ m → pred (sucℤ m) ≡ m
+pred-suc m = begin
+  pred (sucℤ m) ≡⟨ sym (+-assoc (- + 1) (+ 1) m) ⟩
+  + 0 + m       ≡⟨ +-identityˡ m ⟩
+  m ∎ where open ≡-Reasoning
+
++-pred : ∀ m n → m + pred n ≡ pred (m + n)
```
Comment (MatthewDaggitt) [inline src/Data/Integer/Properties.agda]: Hmm the naming and the direction of the equality of these two should probably agree with `neg-distribˡ-*` earlier in the file?

File: src/Data/Integer/Divisibility/Properties.agda [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+------------------------------------------------------------------------
+-- Properties of _∣′_
+
+_∣′?_ : Decidable _∣′_
+k ∣′? m = DEC.map′ ∣m⇒∣′m ∣′m⇒∣m (∣ k ∣ Ndiv.∣? ∣ m ∣)
+
+0∣′⇒≡0 : ∀ {m} → + 0 ∣′ m → m ≡ + 0
+0∣′⇒≡0 0|m = ∣n∣≡0⇒n≡0 (Ndiv.0∣⇒≡0 (∣′m⇒∣m 0|m))
+
+m∣′∣m∣ : ∀ {m} → m ∣′ + ∣ m ∣
```
Comment (MatthewDaggitt) [inline src/Data/Integer/Divisibility/Properties.agda]: The types of these are **really** hard to parse. Might be worth adding a few judicious brackets?

File: src/Data/Integer/Divisibility.agda [hunk 6]
```diff
@@ -17,11 +17,17 @@ open import Relation.Binary.PropositionalEquality
 
 -- Divisibility.
 
-infix 4 _∣_
+infix 4 _∣_ _∣′_
```
Comment (MatthewDaggitt) [inline src/Data/Integer/Divisibility.agda]: Hmm so I think in most places the library uses \prime rather than '?

File: src/Data/Integer/Properties.agda [hunk 7]
```diff
@@ -867,32 +1270,76 @@ n≤1+n n = ≤-step ≤-refl
...
+-- Other properties of _<_
+
 n≮n : ∀ {n} → n ≮ n
 n≮n {+ n}           (+≤+ n<n) =  contradiction n<n ℕₚ.1+n≰n
 n≮n { -[1+ 0 ]}     ()
 n≮n { -[1+ suc n ]} (-≤- n<n) =  contradiction n<n ℕₚ.1+n≰n
 
-<⇒≤ : ∀ {m n} → m < n → m ≤ n
-<⇒≤ m<n =  ≤-trans (n≤1+n _) m<n
+>→≰ : ∀ {x y} → x > y → x ≰ y
```
Comment (MatthewDaggitt) [inline src/Data/Integer/Properties.agda]: This should be a `⇒` arrow in the name?

**Comment by MatthewDaggitt** [comment]
Okay, so I've finally gone through this thoroughly. Been too intimidated before, apologies! Promise to merge it in soon!

File: src/Data/Integer/Properties.agda (hunk 4 above)
Comment (gallais) [reply src/Data/Integer/Properties.agda]: This one follows [`+-suc`](https://agda.github.io/agda-stdlib/Data.Nat.Properties.html#9415).

File: src/Data/Integer/Properties.agda (hunk 7 above)
Comment (gallais) [reply src/Data/Integer/Properties.agda]: Then we should also rename [`≰→>`](https://agda.github.io/agda-stdlib/Data.Integer.Properties.html#29357)

File: src/Data/Integer/Properties.agda (hunk 2 above)
Comment (gallais) [reply src/Data/Integer/Properties.agda]: There are other `-pos` lemmas in the library which is why I was expecting
something like this to exist (and show up when I use `C-c C-z` with _≤_ and "pos") AFAICR.

It doesn't cost us anything to throw it in.

File: src/Data/Integer/Divisibility.agda (hunk 6 above)
Comment (gallais) [reply src/Data/Integer/Divisibility.agda]: Yep. :( I tried to preserve backwards compatibility.

Alternatively we could put all the prime'd stuff into e.g. `Data.Integer.Divisibility.Signed` and call it `_∣_` too.

**Comment by gallais** [comment]
There's no rush: AFAIK only the Presburger solver relies on this PR and I am not
aware of anyone using it.

File: src/Data/Integer/Divisibility.agda (hunk 6 above)
Comment (MatthewDaggitt) [reply src/Data/Integer/Divisibility.agda]: That seems like a better option. I'll do that.

File: src/Data/Integer/Properties.agda [hunk 8]
```diff
@@ -1404,3 +1403,11 @@ Please use neg-involutive instead."
 "Warning: +-⊖-left-cancel was deprecated in v0.15.
 Please use +-cancelˡ-⊖ instead."
 #-}
+
+-- Version 0.18
+
+≰→> = ≰⇒>
+{-# WARNING_ON_USAGE ≰→>
+"Warning: ≰→> was deprecated in v0.15.
```
Comment (gallais) [inline src/Data/Integer/Properties.agda]: Shouldn't this be v0.18?

Comment (MatthewDaggitt) [reply src/Data/Integer/Properties.agda]: Eek, yes. Didn't do a very good job of this!

File: src/Data/Integer/Coprimality.agda [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+------------------------------------------------------------------------
+-- Definition
+
+Coprime : Rel ℤ 0ℓ
+Coprime = ℕ.Coprime on ∣_∣
+
+------------------------------------------------------------------------
+-- Properties of comprimality
+
```
Comment (gallais) [inline src/Data/Integer/Coprimality.agda]: Typo

**Comment by gallais** [comment]
> Maybe you proved it under 2.6.0?

Even with 2.6.0 I now get a termination error. That's because of `--without-K`. Cf. agda/agda#1680
We can fix it by splitting the definition in `pos` and `neg`. Pushing now.

**Comment by gallais** [comment]
> Maybe you proved it under 2.6.0?

Even with 2.6.0 I now get a termination error. That's because of `--without-K`. Cf. agda/agda#1680
We can fix it by splitting the definition in `pos` and `neg`. Pushing now.

**Comment by gallais** [comment]
And with this merge the stdlib is now more than 50k LoC! :tada:

**Comment by MatthewDaggitt** [comment]
Haha, where do you find that out?

**Comment by gallais** [comment]
I sometimes run `find src -name "*.agda" | xargs wc -l` in the projects' root.
